db/migrate/20190123145942_add_type_column_to_service_order.rb

```diff
@@ -0,0 +1,16 @@
+class AddTypeColumnToServiceOrder < ActiveRecord::Migration[5.0]
+  def up
+    add_column :service_order, :type, :string
+
+    say_with_time("Set service_order type") do
+      ServiceOrder.all.each do |service_order|
+        type = service_order.miq_requests.pluck(:type).any? { |t| t === 'ServiceTemplateTransformationPlanRequest' } ? ServiceOrderV2V : ServiceOrderCart
+        service_order.update_attribute(:type, type)
```

Comment:
A few thoughts on these changes.
1. `add_colum` and `remove_column ` reference `service_order` instead of `service_orders` (plural)
1. Use `update` instead of `update_attribute`
1. Since you are manipulating the `type` column directly I would recommend disabling STI at the top of the migration with `self.inheritance_column = :_type_disabled` in the class.
1. I recall we try to avoid active_record relationships in migration and use direct queries.  In this case that would mean avoiding the `service_order.miq_requests` call.
1. There are references to new class types `ServiceOrderV2V : ServiceOrderCart` which are not merged yet and the migration will fail.  (See https://github.com/ManageIQ/manageiq/pull/19766)
1. **PR is missing migration tests.**

Then I really started thinking about the changes and wanted to suggest the following migration replacement.

The basic premise is we get all the v2v Requests and do a single call to update the related ServiceOrders.  Any ServiceOrders remaining will be cart orders, so we can update those by finding ones that have `type => nil`.

Here's what I think the migration should look like:  (
```ruby
class AddTypeColumnToServiceOrder < ActiveRecord::Migration[5.0]
  class ServiceOrder < ActiveRecord::Base
    self.inheritance_column = :_type_disabled
  end

  def up
    add_column :service_orders, :type, :string

    say_with_time("Set service_orders type") do
      v2v_order_ids = ServiceTemplateTransformationPlanRequest.pluck(:service_order_id)
      ServiceOrder.where(:id => v2v_order_ids).update_all(:type => 'ServiceOrderV2V')

      ServiceOrder.where(:type => nil).update_all(:type => 'ServiceOrderCart')
    end
  end

  def down
    remove_column :service_orders, :type
  end
end
```

This still requires that https://github.com/ManageIQ/manageiq/pull/19766 gets merged and I highly recommend tests be added to this PR before merging.

Reply:
Also need to update https://github.com/ManageIQ/manageiq/pull/19766 to address the comment above (https://github.com/ManageIQ/manageiq-schema/pull/328#issuecomment-535016745)

> Well, we actually need a way to guarantee that this column gets set automatically when a new ServiceOrder is created, which would cover it. I think that means a before_save hook.

Reply:
It will also need some logic to make sure the v2v paths create the right kind